src/sim/InitializeTracks.cu

```diff
@@ -0,0 +1,306 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file InitializeTracks.cu
+//---------------------------------------------------------------------------//
+#include "InitializeTracks.hh"
+
+#include <thrust/device_ptr.h>
+#include <thrust/reduce.h>
+#include <thrust/remove.h>
+#include <thrust/scan.h>
+#include <vector>
+#include "base/Atomics.hh"
+#include "base/DeviceVector.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+// KERNELS
+//---------------------------------------------------------------------------//
+/*!
+ * Initialize the track states on device. The track initializers are created
+ * from either primary particles or secondaries. The new tracks are inserted
+ * into empty slots (vacancies) in the track vector.
+ *
+ * TODO: Add sim states
+ */
+__global__ void
+initialize_tracks_kernel(size_type                    num_tracks,
+                         const span<size_type>        vacancies,
+                         const span<TrackInitializer> initializers,
+                         const ParticleParamsPointers pparams,
+                         const ParticleStatePointers  pstates,
+                         const GeoParamsPointers      gparams,
+                         const GeoStatePointers       gstates)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < num_tracks)
+    {
+        // Get the track initializer, starting from the back of the vector
+        size_type               init_id = initializers.size() - thread_id - 1;
```

Comment:
You're right, that `num_tracks` is superfluous. The `vacancies` and `initializers` aren't necessarily the same size; `vacancies` is resized just before this kernel launch so it will have at most the same number of elements as `initializers`, but `initializers` could be larger. Then `n = vacancies.size()` track initializers are pulled off the back of that vector so that the first vacancy is filled with the last track initializer, the second vacancy with the next-to-last, etc. This is all pretty poorly documented (I confused myself looking back at it just now), so I’ll go back and add some more descriptive comments.

Reply:
Hehe there's a lot to keep "track" of. 😉 Thanks!